Python/Include/Scalar.hpp

```diff
@@ -0,0 +1,173 @@
+      PythonScalarConstPointer(const PythonScalar<T>* ptr);
+
+      //! Converts the pointer to const PythonScalar<T>*.
+      operator const PythonScalar<T>*() const;
+
+      //! Converts the pointer to const T*.
+      operator const T*() const;
+
+    private:
+      const PythonScalar<T>* m_ptr;
+  };
+
+  //! Adapts a scalar type to pybind11::object by providing the one-way
+  //! conversion from Python objects and forwarding scalar semantics.
+  /*
+    \tparam T The type of the scalar - an arithmetic type.
+  */
+  template<typename T>
+  class PythonScalar {
+    public:
+
+      //! Constructs an uninitialized PythonScalar.
+      PythonScalar() = default;
+
+      //! Constructs a PythonScalar from a Python object.
+      PythonScalar(const pybind11::object& obj);
+
+      //! Constructs a PythonScalar from an instance of the scalar.
+      PythonScalar(T value);
+
+      //! Converts the PythonScalar to the type of the scalar.
+      operator T() const;
+
+      //! Converts the PythonScalar to a reference to the scalar.
+      operator T&();
+
+      //! Produces a PythonScalarPointer corresponding to the PythonScalar.
+      PythonScalarPointer<T> operator &();
+
+      //! Produces a PythonScalarConstPointer corresponding to the
+      //! PythonScalar.
+      PythonScalarConstPointer<T> operator &() const;
+
+    private:
+      friend PythonScalarPointer<T>;
+      friend PythonScalarConstPointer<T>;
+
+      T m_value;
+  };
+
+  template<typename T>
+  PythonScalarPointer<T>::PythonScalarPointer(PythonScalar<T>* ptr) 
+    : m_ptr(ptr) {}
+
+  template<typename T>
+  PythonScalarPointer<T>::operator PythonScalar<T>*() const {
+    return m_ptr;
+  }
+
+  template<typename T>
+  PythonScalarPointer<T>::operator T*() const {
+    return &(m_ptr->m_value);
+  }
+
+  template<typename T>
+  PythonScalarConstPointer<T>::PythonScalarConstPointer(const PythonScalar<T>*
+      ptr)
+    : m_ptr(ptr) {}
+
+  template<typename T>
+  PythonScalarConstPointer<T>::operator const PythonScalar<T>*() const {
+    return m_ptr;
+  }
+
+  template<typename T>
+  PythonScalarConstPointer<T>::operator const T*() const {
+    return &(m_ptr->m_value);
+  }
+
+  template<typename T>
+  PythonScalar<T>::PythonScalar(const pybind11::object& obj)
+    : m_value(obj.template cast<T>()) {}
+
+  template<typename T>
+  PythonScalar<T>::PythonScalar(T value)
+    : m_value(std::move(value)) {}
+
+  template<typename T>
+  PythonScalarPointer<T> PythonScalar<T>::operator &() {
+    return this;
+  }
+
+  template<typename T>
+  PythonScalarConstPointer<T> PythonScalar<T>::operator &() const {
+    return this;
+  }
+
+  template<typename T>
+  PythonScalar<T>::operator T() const {
+    return m_value;
+  }
+
+  template<typename T>
+  PythonScalar<T>::operator T&() {
+    return m_value;
+  }
+
+  template<typename T>
+  PythonScalar<T> abs(PythonScalar<T> scalar) {
+    return ::abs(static_cast<T>(scalar));
```

Comment:
Hmm... I suspect this won't work for types other than ```int``` and ```double``` and related fundamental types. Is that sufficient? If so that's fine, just want to confirm.

Reply:
Yes, `Scalars `are supposed to be for fundamental types only. They can be custom types as well, but those should be fundamental enough to allow for most built-in operations and `math.h` functions. The reason is that most processing frameworks (as well as our code, if it gets more complicated) will heavily rely on those types being fundamental.
If a (C++) user wants to use a custom scalar type for computation, it is up to them to provide with all the necessary math adaptors.

Reply:
The issue is that there is no ```abs``` function for ```double``` in the global namespace. You would have to use ADL to get this to work which is kind of akward. You will need to do something like:

    template<typename T>
    auto internal_abs(const T& value) {
      using namespace std;
      return abs(value);
    }

And then call ```internal_abs``` from within ```PythonScalar```. Or mark the constructor as explicit to avoid an implicit conversion from ```T``` to ```PythonScalar```.

Reply:
Can't I just move the definition out of namespace Rover? Such as:
```
  template<typename T>
  Rover::PythonScalar<T> abs(Rover::PythonScalar<T> scalar) {
    return abs(static_cast<T>(scalar));
  }
```
Seems to work for me, e.g. in `namespace pybind11`.
On the other hand, I am now not sure why the original code worked if there is no function `abs` in the global namespace.

Reply:
There is an ```abs``` in the global namespace but it is only defined for integral types, so a ```double``` gets implicitly converted to an ```int```.

Reply:
To use the ```abs``` for ```double```, you need ```std::abs```, which means to include it in the overload set you'll need to use ```using namespace std;```.

Reply:
Still strange, as VS has never produced the narrowing conversion warning.
Okay, so if put
```
  template<typename T>
  Rover::PythonScalar<T> abs(Rover::PythonScalar<T> scalar) {
    using namespace std;
    return abs(static_cast<T>(scalar));
  }
```
outside of `namespace Rover`, does it suffice for types `std::abs` exists for? If a custom scalar type is created, the creator is required to supply `abs` in the same namespace as the type itself, hence enabling ADL without `using namespace`? And anyway, it is very unlikely for PythonScalar to be used with custom types, namely anything except `double`, `float`, and maybe `long double` in extreme circumstances. Even if such types are used, `PythonScalar` is a part of Model's export rather than a standalone user-exposed class, so it seems logically fine to require users to conform or even modify the adaptor class itself where necessary.